File: src/main/java/scripting/npc/NPCConversationManager.java
package scripting.npc;

import client.Client;
import field.object.drop.DropEntry;
import field.object.life.FieldMobTemplate;
import lombok.Getter;
import lombok.NonNull;
import lombok.Setter;
import managers.MobManager;
import net.database.DropAPI;
import net.maple.packets.ConversationPackets;
import scripting.AbstractPlayerInteraction;

import java.util.AbstractMap;
import java.util.List;
import java.util.function.Function;

@Getter
@SuppressWarnings("unused")
public class NPCConversationManager extends AbstractPlayerInteraction {

    protected @NonNull final int npcId;
    private @Setter String text = "";

    public NPCConversationManager(@NonNull Client c, int npcId) {
        super(c);
        this.npcId = npcId;
    }

    public void sendOk(String text) {
        c.write(ConversationPackets.getOkMessagePacket(npcId, 0, text));
    }

    public void sendOk(String text, int speaker) {
        c.write(ConversationPackets.getOkMessagePacket(npcId, speaker, text));
    }

    public void sendNext(String text) {
        c.write(ConversationPackets.getNextMessagePacket(npcId, 0, text));
    }

    public void sendNext(String text, int speaker) {
        c.write(ConversationPackets.getNextMessagePacket(npcId, speaker, text));
    }

    public void sendPrev(String text) {
        c.write(ConversationPackets.getPrevMessagePacket(npcId, 0, text));
    }

    public void sendPrev(String text, int speaker) {
        c.write(ConversationPackets.getPrevMessagePacket(npcId, speaker, text));
    }

    public void sendNextPrev(String text) {
        c.write(ConversationPackets.getNextPrevMessagePacket(npcId, 0, text));
    }

    public void sendNextPrev(String text, int speaker) {
        c.write(ConversationPackets.getNextPrevMessagePacket(npcId, speaker, text));
    }

    public void sendYesNo(String text) {
        c.write(ConversationPackets.getYesNoMessagePacket(npcId, 0, text));
    }

    public void sendYesNo(String text, int speaker) {
        c.write(ConversationPackets.getYesNoMessagePacket(npcId, speaker, text));
    }

    public void sendGetText(String text, String def, int min, int max) {
        c.write(ConversationPackets.getTextMessagePacket(npcId, 0, text, def, min, max));
    }

    public void sendGetText(String text, String def, int min, int max, int speaker) {
        c.write(ConversationPackets.getTextMessagePacket(npcId, speaker, text, def, min, max));
    }

    public void sendGetNumber(String text, int def, int min, int max) {
        c.write(ConversationPackets.getNumberMessagePacket(npcId, 0, text, def, min, max));
    }

    public void sendGetNumber(String text, int def, int min, int max, int speaker) {
        c.write(ConversationPackets.getNumberMessagePacket(npcId, speaker, text, def, min, max));
    }

    public void sendSimple(String text) {
        c.write(ConversationPackets.getSimpleMessagePacket(npcId, 0, text));
    }

    public void sendSimple(String text, int speaker) {
        c.write(ConversationPackets.getSimpleMessagePacket(npcId, speaker, text));
    }

    public void sendAcceptDecline(String text) {
        c.write(ConversationPackets.getAcceptMessagePacket(npcId, 0, text));
    }

    public void sendAcceptDecline(String text, int speaker) {
        c.write(ConversationPackets.getAcceptMessagePacket(npcId, speaker, text));
    }

    public void sendGetTextBox() {
        c.write(ConversationPackets.getBoxTextMessagePacket(npcId, 0, "", 48, 6));
    }

    public void sendGetTextBox(int speaker) {
        c.write(ConversationPackets.getBoxTextMessagePacket(npcId, speaker, "", 48, 6));
    }

    public void sendGetTextBox(String def, int cols, int rows) {
        c.write(ConversationPackets.getBoxTextMessagePacket(npcId, 0, def, cols, rows));
    }

    public void sendGetTextBox(String def, int cols, int rows, int speaker) {
        c.write(ConversationPackets.getBoxTextMessagePacket(npcId, speaker, def, cols, rows));
    }

    public void sendSlide(String text, int type, int selected) {
        c.write(ConversationPackets.getSlideMenuMessagePacket(npcId, 0, text, type, selected));
    }

    public void sendSlide(String text, int type, int selected, int speaker) {
        c.write(ConversationPackets.getSlideMenuMessagePacket(npcId, speaker, text, type, selected));
    }

    public void dispose() {
        NPCScriptManager.getInstance().dispose(this);
    }

    public void startQuest(int qid) {
        getPlayer().startQuest(qid, npcId);
    }

    public void completeQuest(int qid) {
        getPlayer().completeQuest(qid);
    }

    /**
     * cm.letters("Hello world"); will show Christmas letters Hello world
     *
     * @param input the text to turn into christmas :)
     * @return String with item images
     */
    public String letters(String input) {
        StringBuilder str = new StringBuilder();
        for (int i = 0; i < input.length(); i++) {
            if (input.charAt(i) == ' ') {
                str.append("\t");
            } else {
                str.append("#i").append(convert(input.charAt(i))).append("#");
            }
        }
        return str.toString();
    }

    private int convert(char in) {
        int upper = 3991000;
        int lower = 3991026;
        int output = Character.isUpperCase(in) ? upper : lower;
        switch (Character.toLowerCase(in)) {
            case 'a':
                output += 0;
                break;
            case 'b':
                output += 1;
                break;
            case 'c':
                output += 2;
                break;
            case 'd':
                output += 3;
                break;
            case 'e':
                output += 4;
                break;
            case 'f':
                output += 5;
                break;
            case 'g':
                output += 6;
                break;
            case 'h':
                output += 7;
                break;
            case 'i':
                output += 8;
                break;
            case 'j':
                output += 9;
                break;
            case 'k':
                output += 10;
                break;
            case 'l':
                output += 11;
                break;
            case 'm':
                output += 12;
                break;
            case 'n':
                output += 13;
                break;
            case 'o':
                output += 14;
                break;
            case 'p':
                output += 15;
                break;
            case 'q':
                output += 16;
                break;
            case 'r':
                output += 17;
                break;
            case 's':
                output += 18;
                break;
            case 't':
                output += 19;
                break;
            case 'u':
                output += 20;
                break;
            case 'v':
                output += 21;
                break;
            case 'w':
                output += 22;
                break;
            case 'x':
                output += 23;
                break;
            case 'y':
                output += 24;
                break;
            case 'z':
                output += 25;
                break;
        }
        return output;
    }

    public List<DropEntry> getMobDrops(int id) {
        FieldMobTemplate template = MobManager.getMob(id);
        if (template != null) {
            if (template.getDrops() == null) {
                template.setDrops(DropAPI.getMobDrops(template.getId()));
            }
            return template.getDrops();
        }
        return null;
    }

    public void addMobDrop(int mid, int iid, double chance) {
        addMobDrop(mid, iid, 1, 1, 0, chance);
    }

    public void addMobDrop(int mid, int iid, int min, int max, double chance) {
        addMobDrop(mid, iid, min, max, 0, chance);
    }

    public void addMobDrop(int mid, int iid, int min, int max, int questid, double chance) {
        DropAPI.addMobDrop(mid, iid, min, max, questid, chance);
    }

    public void editDropChance(int mid, int iid, double chance) {
        DropAPI.updateDropChance(mid, iid, chance);
    }

    public void removeDrop(int mid, int iid) {
        DropAPI.removeDrop(mid, iid);
    }

    public void editMinMaxChance(int mid, int iid, int min, int max, double chance) {
        DropAPI.updateMinMaxChance(mid, iid, min, max, chance);
    }
}

File: src/main/java/scripting/npc/NPCConversationManager.kt
package scripting.npc

import client.Client
import client.player.Beauty
import field.obj.drop.DropEntry
import field.obj.life.FieldMobTemplate
import managers.BeautyManager
import managers.MobManager
import net.database.BeautyAPI
import net.database.DropAPI
import net.database.DropAPI.updateDropChance
import net.database.DropAPI.updateMinMaxChance
import net.maple.packets.ConversationPackets
import scripting.AbstractPlayerInteraction
import java.util.*
import java.util.function.Consumer
import java.util.stream.Collectors
import kotlin.collections.ArrayList
import kotlin.collections.LinkedHashMap
import kotlin.collections.set

open class NPCConversationManager(c: Client, val npcId: Int) : AbstractPlayerInteraction(c) {

    var text = ""

    fun sendOk(text: String) {
        c.write(ConversationPackets.getOkMessagePacket(npcId, 0, text))
    }

    fun sendOk(text: String, speaker: Int) {
        c.write(ConversationPackets.getOkMessagePacket(npcId, speaker, text))
    }

    fun sendNext(text: String) {
        c.write(ConversationPackets.getNextMessagePacket(npcId, 0, text))
    }

    fun sendNext(text: String, speaker: Int) {
        c.write(ConversationPackets.getNextMessagePacket(npcId, speaker, text))
    }

    fun sendPrev(text: String) {
        c.write(ConversationPackets.getPrevMessagePacket(npcId, 0, text))
    }

    fun sendPrev(text: String, speaker: Int) {
        c.write(ConversationPackets.getPrevMessagePacket(npcId, speaker, text))
    }

    fun sendNextPrev(text: String) {
        c.write(ConversationPackets.getNextPrevMessagePacket(npcId, 0, text))
    }

    fun sendNextPrev(text: String, speaker: Int) {
        c.write(ConversationPackets.getNextPrevMessagePacket(npcId, speaker, text))
    }

    fun sendYesNo(text: String) {
        c.write(ConversationPackets.getYesNoMessagePacket(npcId, 0, text))
    }

    fun sendYesNo(text: String, speaker: Int) {
        c.write(ConversationPackets.getYesNoMessagePacket(npcId, speaker, text))
    }

    fun sendGetText(text: String, def: String, min: Int, max: Int) {
        c.write(ConversationPackets.getTextMessagePacket(npcId, 0, text, def, min, max))
    }

    fun sendGetText(text: String, def: String, min: Int, max: Int, speaker: Int) {
        c.write(ConversationPackets.getTextMessagePacket(npcId, speaker, text, def, min, max))
    }

    fun sendGetNumber(text: String, def: Int, min: Int, max: Int) {
        c.write(ConversationPackets.getNumberMessagePacket(npcId, 0, text, def, min, max))
    }

    fun sendGetNumber(text: String, def: Int, min: Int, max: Int, speaker: Int) {
        c.write(ConversationPackets.getNumberMessagePacket(npcId, speaker, text, def, min, max))
    }

    fun sendSimple(text: String) {
        c.write(ConversationPackets.getSimpleMessagePacket(npcId, 0, text))
    }

    fun sendSimple(text: String, speaker: Int) {
        c.write(ConversationPackets.getSimpleMessagePacket(npcId, speaker, text))
    }

    fun sendAcceptDecline(text: String) {
        c.write(ConversationPackets.getAcceptMessagePacket(npcId, 0, text))
    }

    fun sendAcceptDecline(text: String, speaker: Int) {
        c.write(ConversationPackets.getAcceptMessagePacket(npcId, speaker, text))
    }

    fun sendGetTextBox() {
        c.write(ConversationPackets.getBoxTextMessagePacket(npcId, 0, "", 48, 6))
    }

    fun sendGetTextBox(speaker: Int) {
        c.write(ConversationPackets.getBoxTextMessagePacket(npcId, speaker, "", 48, 6))
    }

    fun sendGetTextBox(def: String, cols: Int, rows: Int) {
        c.write(ConversationPackets.getBoxTextMessagePacket(npcId, 0, def, cols, rows))
    }

    fun sendGetTextBox(def: String, cols: Int, rows: Int, speaker: Int) {
        c.write(ConversationPackets.getBoxTextMessagePacket(npcId, speaker, def, cols, rows))
    }

    fun sendSlide(text: String, type: Int, selected: Int) {
        c.write(ConversationPackets.getSlideMenuMessagePacket(npcId, 0, text, type, selected))
    }

    fun sendSlide(text: String, type: Int, selected: Int, speaker: Int) {
        c.write(ConversationPackets.getSlideMenuMessagePacket(npcId, speaker, text, type, selected))
    }

    open fun dispose() {
        NPCScriptManager.dispose(this)
    }

    fun startQuest(qid: Int) {
        player.startQuest(qid, npcId)
    }

    fun completeQuest(qid: Int) {
        player.completeQuest(qid)
    }

    /**
     * cm.letters("Hello world"); will show Christmas letters Hello world
     *
     * @param input the text to turn into christmas :)
     * @return String with item images
     */
    fun letters(input: String): String {
        val str = StringBuilder()
        for (i in input.indices) {
            if (input[i] == ' ') {
                str.append("\t")
            } else {
                str.append("#i").append(convert(input[i])).append("#")
            }
        }
        return str.toString()
    }

    private fun convert(input: Char): Int {
        val upper = 3991000
        val lower = 3991026
        var output = if (Character.isUpperCase(input)) upper else lower
        when (Character.toLowerCase(input)) {
            'a' -> output += 0
            'b' -> output += 1
            'c' -> output += 2
            'd' -> output += 3
            'e' -> output += 4
            'f' -> output += 5
            'g' -> output += 6
            'h' -> output += 7
            'i' -> output += 8
            'j' -> output += 9
            'k' -> output += 10
            'l' -> output += 11
            'm' -> output += 12
            'n' -> output += 13
            'o' -> output += 14
            'p' -> output += 15
            'q' -> output += 16
            'r' -> output += 17
            's' -> output += 18
            't' -> output += 19
            'u' -> output += 20
            'v' -> output += 21
            'w' -> output += 22
            'x' -> output += 23
            'y' -> output += 24
            'z' -> output += 25
        }
        return output
    }

    fun getMobDrops(id: Int): List<DropEntry> {
        val template: FieldMobTemplate = MobManager.getMob(id) ?: return ArrayList()
        if (template.drops == null) {
            template.drops = DropAPI.getMobDrops(template.id)
        }
        return template.drops?.toList() ?: return ArrayList()
    }

    fun addMobDrop(mid: Int, iid: Int, chance: Double) {
        addMobDrop(mid, iid, 1, 1, 0, chance)
    }

    fun addMobDrop(mid: Int, iid: Int, min: Int, max: Int, chance: Double) {
        addMobDrop(mid, iid, min, max, 0, chance)
    }

    fun addMobDrop(mid: Int, iid: Int, min: Int, max: Int, questid: Int, chance: Double) {
        DropAPI.addMobDrop(mid, iid, min, max, questid, chance)
    }

    fun editDropChance(mid: Int, iid: Int, chance: Double) {
        updateDropChance(mid, iid, chance)
    }

    fun removeDrop(mid: Int, iid: Int) {
        DropAPI.removeDrop(mid, iid)
    }

    fun editMinMaxChance(mid: Int, iid: Int, min: Int, max: Int, chance: Double) {
        updateMinMaxChance(mid, iid, min, max, chance)
    }

    fun openStorage() {
        openStorage(npcId)
    }

    fun postRewards(rewards: Any) {
        val map: AbstractMap<*, *> = rewards as AbstractMap<*, *>
        val counter = 1
        val isNull = false
        val exp = map["exp"]
        val mesos = map["mesos"]
        val fame = map["fame"]
        val random = map["random"]
        val items = map["items"] as AbstractMap<*, *>?
        val readableItemMap: MutableMap<Int, Int> = LinkedHashMap()
        items!!.values.forEach(Consumer { pair: Any -> readableItemMap[(pair as AbstractMap<*, *>)["0"] as Int] = pair["1"] as Int })
        println(readableItemMap)
    }

    val allHairs: List<Beauty> get() = ArrayList(BeautyManager.getHairs().values)

    fun getEnabledHairs(gender: Int): List<Beauty> {
        return BeautyManager.getHairs().values.stream()
                .filter(Beauty::isEnabled)
                .filter { it.gender == gender }
                .collect(Collectors.toList())
    }

    fun getDisabledHairs(gender: Int): List<Beauty> {
        return BeautyManager.getHairs().values.stream()
                .filter { !it.isEnabled }
                .filter { it.gender == gender }
                .collect(Collectors.toList())
    }

    fun updateHair(id: Int) {
        val b = BeautyManager.getHairs()[id] ?: return
        b.isEnabled = !b.isEnabled
        BeautyAPI.updateHair(id)
    }
}